src/main/java/build/buildfarm/instance/server/NodeInstance.java

```diff
@@ -647,10 +652,33 @@ private static void downloadUrl(
       throw new IOException(message);
     }
 
-    try (InputStream in = connection.getInputStream();
-        OutputStream out = getContentOutputStream.create(contentLength)) {
+    IOSupplier<InputStream> inSupplier;
+    if (expectedHash.isEmpty() || contentLength < 0) {
+      // not great, plenty risky for large objects
```

Comment:
Should there be a cap on this inputStream? like, abort after some value? At least some logging. I'd hate to try to pull a (large) OCI layer through this path 😬

Reply:
There's not currently any storage that would permit us to flow through to mitigate heap usage configured on servers.
I'd be fine with this, or another mechanism that cuts it off and decides that we need to redownload for the actual write. More engineering required.